dropshot/src/router.rs

```diff
@@ -464,9 +481,34 @@ impl<'a, Context: ServerContext> Iterator for HttpRouterIter<'a, Context> {
 
 /**
  * Helper function for taking a Uri path and producing a `Vec<String>` of
- * URL-encoded strings, each representing one segment of the path.
+ * URL-decoded strings, each representing one segment of the path. The input is
+ * percent-encoded. Empty segments i.e. due to consecutive "/" characters or a
+ * leading "/" are omitted.
+ *
+ * Regarding "dot-segments" ("." and ".."), RFC 3986 section 3.3 says this:
+ *    The path segments "." and "..", also known as dot-segments, are
+ *    defined for relative reference within the path name hierarchy.  They
+ *    are intended for use at the beginning of a relative-path reference
+ *    (Section 4.2) to indicate relative position within the hierarchical
+ *    tree of names.  This is similar to their role within some operating
+ *    systems' file directory structures to indicate the current directory
+ *    and parent directory, respectively.  However, unlike in a file
+ *    system, these dot-segments are only interpreted within the URI path
+ *    hierarchy and are removed as part of the resolution process (Section
+ *    5.2).
+ *
+ * As dropshot requests are necessarily made with absolute (rather than
+ * relative) targets (any resolution presumed to have been done by the client)
+ * we make no special interpretation of dot-segments. In short, the strings "."
+ * and ".." may appear as segments (and thus as the values for path-derived
+ * variables). Consumers should take special care concerning these values when
+ * using Dropshot to serve files. Absent proper input validation, it could be
+ * possible for requests to retrieve files not-intended to be exposed (e.g.
+ * "GET /../../../etc/passwd").
```

Comment:
Given the position that "clients are responsible for correctly canonicalizing . and ..", I'm actually kinda surprised with the conclusion that "Dropshot will ignore all that stuff, and hope the client got it right".

Couldn't we explicitly ban "." and ".." in path segments? If we throw an error on their discovery, we'd at least halt "known-bad-client behavior", with minimal effort (since we can still hold the stance "Dropshot won't resolve it for you").

Reply:
I spoke at length with @davepacheco about this. I'm not sure why we didn't arrive at your suggestion--perhaps we had been too micro-focused on various RFCs.

If a client were to implement file access they would need to consider symlinks as well as dot segments--that would obviously be well beyond what Dropshot itself could do.

All that said: I agree with your position and I don't think it's particularly valuable to allow an API to specify ".." as a path segment for an endpoint. @davepacheco if you disagree with this it seems easy to undo later.

Reply:
Cool, sounds good to me! I figured it would be better to start with a "safe-but-conservative" approach, since we can always walk it back later (but the other direction would be more difficult).